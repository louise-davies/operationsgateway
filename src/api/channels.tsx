import React from 'react';
import axios, { AxiosError } from 'axios';
import {
  FullChannelMetadata,
  FullScalarChannelMetadata,
  isChannelMetadataScalar,
  timeChannelName,
} from '../app.types';
import {
  useQuery,
  UseQueryResult,
  UseQueryOptions,
} from '@tanstack/react-query';
import { Column } from 'react-table';
import {
  renderImage,
  roundNumber,
} from '../table/cellRenderers/cellContentRenderers';
import { selectUrls } from '../state/slices/configSlice';
import { useAppSelector } from '../state/hooks';
import { readSciGatewayToken } from '../parseTokens';

interface ChannelsEndpoint {
  channels: {
    [systemName: string]: Omit<FullChannelMetadata, 'systemName'>;
  };
}

// This metadata is always present in every record
export const staticChannels: { [systemName: string]: FullChannelMetadata } = {
  [timeChannelName]: {
    systemName: timeChannelName,
    name: 'Time',
    type: 'scalar',
    path: '/system',
  },
  shotnum: {
    systemName: 'shotnum',
    name: 'Shot Number',
    type: 'scalar',
    path: '/system',
  },
  activeArea: {
    systemName: 'activeArea',
    name: 'Active Area',
    type: 'scalar',
    path: '/system',
  },
  activeExperiment: {
    systemName: 'activeExperiment',
    name: 'Active Experiment',
    type: 'scalar',
    path: '/system',
  },
};

const fetchChannels = (apiUrl: string): Promise<FullChannelMetadata[]> => {
  return axios
    .get<ChannelsEndpoint>(`${apiUrl}/channels`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      const { channels } = response.data;

      if (!channels || Object.keys(channels).length === 0) return [];

      const convertedChannels: FullChannelMetadata[] = Object.entries(
        channels
      ).map(([systemName, channel]) => ({
        systemName,
        ...channel,
      }));

      return [...Object.values(staticChannels), ...convertedChannels];
    });
};

export interface ChannelSummary {
  first_date: string;
  most_recent_date: string;
  recent_sample: { [timestamp: string]: string | number }[];
}

const fetchChannelSummary = (
  apiUrl: string,
  channel: string
): Promise<ChannelSummary> => {
  return axios
    .get(`${apiUrl}/channels/summary/${channel}`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data;
    });
};

// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-constraint
export const useChannels = <T extends unknown = FullChannelMetadata[]>(
  options?: UseQueryOptions<FullChannelMetadata[], AxiosError, T, string[]>
): UseQueryResult<T, AxiosError> => {
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery(
    ['channels'],
    (params) => {
      return fetchChannels(apiUrl);
    },
    {
      onError: (error) => {
        console.log('Got error ' + error.message);
      },
      ...(options ?? {}),
    }
  );
};

export const useChannelSummary = (
  channel: string | undefined
): UseQueryResult<ChannelSummary, AxiosError> => {
  const { apiUrl } = useAppSelector(selectUrls);
  const dataChannel =
    typeof channel !== 'undefined' && !(channel in staticChannels)
      ? channel
      : '';

  return useQuery(
    ['channelSummary', dataChannel],
    (params) => {
      return fetchChannelSummary(apiUrl, dataChannel);
    },
    {
      onError: (error) => {
        console.log('Got error ' + error.message);
      },
      enabled: dataChannel.length !== 0,
    }
  );
};

export const constructColumns = (channels: FullChannelMetadata[]): Column[] => {
  const myColumns: Column[] = [];

  channels.forEach((channel: FullChannelMetadata) => {
    const newColumn: Column = {
      Header: () => {
        const headerName = channel.name ? channel.name : channel.systemName;
        // Provide an actual header here when we have it
        // TODO: do we need to split on things other than underscore?
        const parts = headerName.split('_');
        const wordWrap = parts.map(
          (part, i) =>
            // \u200B renders a zero-width space character
            // which allows line-break but isn't visible
            part + (i < parts.length - 1 ? '_\u200B' : '')
        );
        return <React.Fragment>{wordWrap.join('')}</React.Fragment>;
      },
      accessor: channel.systemName,
      // TODO: get these from data channel info
      channelInfo: channel,
    };
    if (isChannelMetadataScalar(channel)) {
      newColumn.Cell = ({ value }) =>
        typeof value === 'number' && typeof channel.precision === 'number' ? (
          <React.Fragment>
            {roundNumber(value, channel.precision, channel.notation)}
          </React.Fragment>
        ) : (
          <React.Fragment>{String(value ?? '')}</React.Fragment>
        );
    } else {
      newColumn.Cell = ({ row, value }) =>
        renderImage(
          value,
          `${channel.name ?? channel.systemName} ${
            channel.type
          } for timestamp ${row.values[timeChannelName]}`
        );
    }
    myColumns.push(newColumn);
  });
  return myColumns;
};

export const getScalarChannels = (
  channels: FullChannelMetadata[]
): FullScalarChannelMetadata[] => {
  return channels.filter(
    (channel) => channel.type === 'scalar'
  ) as FullScalarChannelMetadata[];
};

const useScalarChannelsOptions = {
  select: (data: FullChannelMetadata[]) => getScalarChannels(data),
};

const useAvailableColumnsOptions = {
  select: (data: FullChannelMetadata[]) => constructColumns(data),
};

export const useScalarChannels = (): UseQueryResult<
  FullScalarChannelMetadata[],
  AxiosError
> => {
  return useChannels(useScalarChannelsOptions);
};

export const useAvailableColumns = (): UseQueryResult<Column[], AxiosError> => {
  return useChannels(useAvailableColumnsOptions);
};
