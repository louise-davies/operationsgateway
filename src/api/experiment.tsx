import { useQuery, UseQueryResult } from '@tanstack/react-query';
import axios, { AxiosError } from 'axios';
import { ExperimentParams } from '../app.types';
import { readSciGatewayToken } from '../parseTokens';
import { useAppSelector } from '../state/hooks';
import { selectUrls } from '../state/slices/configSlice';

const fetchExperiment = (apiUrl: string): Promise<ExperimentParams[]> => {
  return axios
    .get(`${apiUrl}/experiments`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data;
    });
};

export const useExperiment = (): UseQueryResult<
  ExperimentParams[],
  AxiosError
> => {
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery({
    queryKey: ['experiments'],

    queryFn: () => fetchExperiment(apiUrl),
  });
};
