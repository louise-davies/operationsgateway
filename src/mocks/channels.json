{
  "_id": "123",
  "channels": {
    "CHANNEL_ABCDE": {
      "name": "Channel_ABCDE",
      "path": "/Channels/1/ABCDE",
      "type": "scalar"
    },
    "CHANNEL_BCDEF": {
      "name": "Channel_BCDEF",
      "path": "/Channels/1/BCDEF",
      "type": "image"
    },
    "CHANNEL_CDEFG": {
      "name": "Channel_CDEFG",
      "path": "/Channels/1/CDEFG",
      "type": "waveform",
      "x_units": "X",
      "y_units": "Y"
    },
    "CHANNEL_DEFGH": {
      "name": "Channel_DEFGH",
      "path": "/Channels/2/DEFGH",
      "type": "scalar",
      "notation": "scientific",
      "precision": 1
    },
    "CHANNEL_EFGHI": {
      "name": "Channel_EFGHI",
      "path": "/Channels/2/EFGHI",
      "type": "image"
    },
    "CHANNEL_FGHIJ": {
      "name": "Channel_FGHIJ",
      "path": "/Channels/2/FGHIJ",
      "type": "waveform"
    },
    "CHANNEL_GHIJK": {
      "name": "Channel_GHIJK",
      "path": "/Channels/3/GHIJK",
      "type": "scalar"
    }
  }
}
