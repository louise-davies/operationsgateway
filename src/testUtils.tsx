import { Action, ThunkAction } from '@reduxjs/toolkit';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import type { RenderOptions } from '@testing-library/react';
import { render } from '@testing-library/react';
import { matchRequestUrl } from 'msw';
import React from 'react';
// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
// need to mock <canvas> for plotting
import 'jest-canvas-mock';
import { Provider } from 'react-redux';
import { staticChannels } from './api/channels';
import {
  DEFAULT_WINDOW_VARS,
  FullChannelMetadata,
  FullScalarChannelMetadata,
  PlotDataset,
  timeChannelName,
} from './app.types';
import channelsJson from './mocks/channels.json';
import { server } from './mocks/server';
import { COLOUR_ORDER } from './plotting/plotSettings/colourGenerator';
import { initialState as initialConfigState } from './state/slices/configSlice';
import { initialState as initialFilterState } from './state/slices/filterSlice';
import { initialState as initialFunctionsState } from './state/slices/functionsSlice';
import {
  PlotConfig,
  initialState as initialPlotState,
} from './state/slices/plotSlice';
import { initialStateFunc as initialSearchStateFunc } from './state/slices/searchSlice';
import { initialState as initialSelectionState } from './state/slices/selectionSlice';
import { initialState as initialTableState } from './state/slices/tableSlice';
import { initialState as initialWindowsState } from './state/slices/windowSlice';
import { AppStore, RootState, setupStore } from './state/store';
import {
  WindowContext,
  WindowContextProvider,
  WindowsRefType,
} from './windows/windowContext';

/**
 * Waits for msw request -
 * @param method string representing the HTTP method
 * @param url string representing the URL match for the route
 * @returns a promise of the matching request
 *  */
export function waitForRequest(method: string, url: string) {
  let newRequestId = '';

  return new Promise<Request>((resolve, reject) => {
    const onRequestStart = ({
      request,
      requestId,
    }: {
      request: Request;
      requestId: string;
    }) => {
      const requestURL = new URL(request.url);
      const matchesMethod =
        request.method.toLowerCase() === method.toLowerCase();

      const matchesUrl = matchRequestUrl(requestURL, url).matches;

      if (matchesMethod && matchesUrl) {
        newRequestId = requestId;
      }
    };

    const onRequestMatch = ({
      request,
      requestId,
    }: {
      request: Request;
      requestId: string;
    }) => {
      if (requestId === newRequestId) {
        server.events.removeListener('request:start', onRequestStart);
        server.events.removeListener('request:match', onRequestMatch);
        server.events.removeListener('request:unhandled', onRequestUnhandled);
        resolve(request);
      }
    };

    const onRequestUnhandled = ({
      request,
      requestId,
    }: {
      request: Request;
      requestId: string;
    }) => {
      const requestURL = new URL(request.url);
      if (requestId === newRequestId) {
        server.events.removeListener('request:start', onRequestStart);
        server.events.removeListener('request:match', onRequestMatch);
        server.events.removeListener('request:unhandled', onRequestUnhandled);
        reject(
          new Error(
            `The ${request.method} ${requestURL.href} request was unhandled.`
          )
        );
      }
    };

    server.events.on('request:start', onRequestStart);

    server.events.on('request:match', onRequestMatch);

    server.events.on('request:unhandled', onRequestUnhandled);
  });
}

export let actions: Action[] = [];
export const resetActions = (): void => {
  actions = [];
};
export const getInitialState = (): RootState => ({
  config: initialConfigState,
  table: initialTableState,
  search: initialSearchStateFunc(),
  plots: initialPlotState,
  filter: initialFilterState,
  functions: initialFunctionsState,
  windows: initialWindowsState,
  selection: initialSelectionState,
});
export const dispatch = (
  action: Action | ThunkAction<void, RootState, unknown, Action<string>>
): void | Promise<void> => {
  if (typeof action === 'function') {
    action(dispatch, getInitialState, null);
    return Promise.resolve();
  } else {
    actions.push(action);
  }
};

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
  queryClient?: QueryClient;
  windowsRef?: React.MutableRefObject<WindowsRefType>;
}

export const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        staleTime: 300000,
      },
    },
  });

export const hooksWrapperWithProviders = (
  state = {},
  queryClient?: QueryClient
) => {
  const testQueryClient = queryClient ?? createTestQueryClient();
  const store = setupStore(state);
  const wrapper = ({ children }: { children: React.ReactNode }) => (
    <Provider store={store}>
      <QueryClientProvider client={testQueryClient}>
        {children}
      </QueryClientProvider>
    </Provider>
  );
  return wrapper;
};

export function renderComponentWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    // Automatically create a query client instance if no query client was passed in
    queryClient = createTestQueryClient(),
    windowsRef,
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element {
    const WindowsProvider = windowsRef
      ? WindowContext.Provider
      : WindowContextProvider;
    return (
      <Provider store={store}>
        <QueryClientProvider client={queryClient}>
          <WindowsProvider {...(windowsRef ? { value: windowsRef } : {})}>
            {children}
          </WindowsProvider>
        </QueryClientProvider>
      </Provider>
    );
  }
  return {
    store,
    queryClient,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}

export function renderComponentWithStore(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderComponentWithStoreAndWindows(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    windowsRef,
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element {
    const WindowsProvider = windowsRef
      ? WindowContext.Provider
      : WindowContextProvider;
    if (windowsRef && windowsRef.current === null) windowsRef.current = {};
    return (
      <Provider store={store}>
        <WindowsProvider {...(windowsRef ? { value: windowsRef } : {})}>
          {children}
        </WindowsProvider>
      </Provider>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export const flushPromises = (): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve));

export const testChannels = [
  ...Object.values(staticChannels),
  ...Object.entries(channelsJson.channels).map(
    ([systemName, channel]) =>
      ({
        systemName,
        ...channel,
      }) as FullChannelMetadata
  ),
];

export const testScalarChannels: FullScalarChannelMetadata[] = [
  ...Object.values(staticChannels).filter(
    (channel) => channel.type === 'scalar'
  ),
  ...Object.entries(channelsJson.channels)
    .filter(([_systemName, channel]) => channel.type === 'scalar')
    .map(
      ([systemName, channel]) =>
        ({
          systemName,
          ...channel,
        }) as FullScalarChannelMetadata
    ),
];

export const generatePlotDataset = (num: number) => {
  const datasetName = testScalarChannels[num].systemName;
  const plotDataset: PlotDataset = {
    name: datasetName,
    data: [
      {
        timestamp: num,
        [datasetName]: num + num,
      },
      {
        timestamp: num + num,
        [datasetName]: num + num + num,
      },
      {
        timestamp: num + num + num,
        [datasetName]: num + num + num + num,
      },
    ],
  };
  return plotDataset;
};

export const testPlotDatasets = Array.from(Array(3), (_, i) =>
  generatePlotDataset(i + 1)
);

export const generatePlotConfig = (num: number) => {
  const plotTitle = `Plot ${num}`;

  const plotConfig: PlotConfig = {
    id: `test-plot-id-${num}`,
    open: num % 2 === 0,
    title: plotTitle,
    plotType: num % 2 === 0 ? 'scatter' : 'line',
    XAxis: num % 3 === 0 ? timeChannelName : undefined,
    XAxisScale:
      num % 3 === 0 ? 'time' : num % 3 === 1 ? 'linear' : 'logarithmic',
    selectedPlotChannels: [],
    leftYAxisScale: num % 2 === 0 ? 'linear' : 'logarithmic',
    rightYAxisScale: num % 2 === 0 ? 'logarithmic' : 'linear',
    gridVisible: num % 2 === 0,
    axesLabelsVisible: num % 2 !== 0,
    selectedColours: [],
    remainingColours: COLOUR_ORDER.map((colour) => colour),
    ...DEFAULT_WINDOW_VARS,
  };

  return plotConfig;
};

export const testPlotConfigs = Array.from(Array(3), (_, i) =>
  generatePlotConfig(i)
);
