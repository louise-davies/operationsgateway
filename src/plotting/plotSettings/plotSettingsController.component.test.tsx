import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { staticChannels } from '../../api/channels';
import { FullScalarChannelMetadata } from '../../app.types';
import { testScalarChannels } from '../../testUtils';
import type { PlotSettingsControllerProps } from './plotSettingsController.component';

describe('Plot Settings component', () => {
  let props: PlotSettingsControllerProps;
  let user: ReturnType<typeof userEvent.setup>;
  const changePlotType = vi.fn();
  const changeXAxis = vi.fn();
  const changeXAxisScale = vi.fn();
  const changeLeftYAxisScale = vi.fn();
  const changeRightYAxisScale = vi.fn();
  const changeSelectedPlotChannels = vi.fn();
  const changeXMinimum = vi.fn();
  const changeXMaximum = vi.fn();
  const changeLeftYAxisMinimum = vi.fn();
  const changeLeftYAxisMaximum = vi.fn();
  const changeRightYAxisMinimum = vi.fn();
  const changeRightYAxisMaximum = vi.fn();
  const changePlotTitle = vi.fn();

  const createView = () => {
    // need to import like this in order for the doMock's to work
    const PlotSettingsController =
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      require('./plotSettingsController.component').default;
    return render(<PlotSettingsController {...props} />);
  };

  beforeEach(() => {
    props = {
      selectedRecordTableChannels: [
        staticChannels['timestamp'] as FullScalarChannelMetadata,
      ],
      allChannels: testScalarChannels,
      changePlotTitle,
      plotType: 'scatter',
      changePlotType,
      XAxis: 'test x-axis',
      changeXAxis,
      XAxisScale: 'linear',
      changeXAxisScale,
      leftYAxisScale: 'logarithmic',
      rightYAxisScale: 'linear',
      changeLeftYAxisScale,
      changeRightYAxisScale,
      leftYAxisLabel: 'left y axis label',
      rightYAxisLabel: 'right y axis label',
      changeLeftYAxisLabel: vi.fn(),
      changeRightYAxisLabel: vi.fn(),
      selectedPlotChannels: [],
      changeSelectedPlotChannels,
      changeXMinimum,
      changeXMaximum,
      changeLeftYAxisMinimum,
      changeLeftYAxisMaximum,
      changeRightYAxisMinimum,
      changeRightYAxisMaximum,
    };

    user = userEvent.setup({ delay: null });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  describe('snapshots', () => {
    beforeAll(() => {
      jest.resetModules();
      jest.doMock('./plotSettingsTextField.component', () => (props) => (
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        <mock-plotSettingsTextField data-testid="mock-plotSettingsTextField">
          {Object.entries(props).map(
            ([propName, propValue]) =>
              `${propName}=${JSON.stringify(propValue, null, 2)}\n`
          )}
          {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
          {/* @ts-ignore */}
        </mock-plotSettingsTextField>
      ));

      jest.doMock('./chartTypeButtons.component', () => (props) => (
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        <mock-chartTypeButtons data-testid="mock-chartTypeButtons">
          {Object.entries(props).map(
            ([propName, propValue]) =>
              `${propName}=${JSON.stringify(propValue, null, 2)}\n`
          )}
          {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
          {/* @ts-ignore */}
        </mock-chartTypeButtons>
      ));

      jest.doMock('./xAxisTab.component', () => (props) => (
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        <mock-xAxisTab data-testid="mock-xAxisTab">
          {Object.entries(props).map(
            ([propName, propValue]) =>
              `${propName}=${JSON.stringify(propValue, null, 2)}\n`
          )}
          {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
          {/* @ts-ignore */}
        </mock-xAxisTab>
      ));

      jest.doMock('./yAxisTab.component', () => (props) => (
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        <mock-yAxisTab data-testid="mock-yAxisTab">
          {Object.entries(props).map(
            ([propName, propValue]) =>
              `${propName}=${JSON.stringify(propValue, null, 2)}\n`
          )}
          {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
          {/* @ts-ignore */}
        </mock-yAxisTab>
      ));
    });

    afterAll(() => {
      jest.resetModules();
      jest.dontMock('./plotSettingsTextField.component');
      jest.dontMock('./chartTypeButtons.component');
      jest.dontMock('./xAxisTab.component');
      jest.dontMock('./yAxisTab.component');
      jest.resetModules();
    });

    it('renders plot settings form correctly (timeseries plot)', () => {
      props.XAxis = 'timestamp';
      props.XAxisScale = 'time';
      const view = createView();

      expect(view.asFragment()).toMatchSnapshot();
    });

    it('renders plot settings form correctly (x-axis tab selected)', () => {
      const view = createView();

      expect(view.asFragment()).toMatchSnapshot();
    });

    it('renders plot settings form correctly (y-axis tab selected)', async () => {
      const view = createView();
      await user.click(screen.getByRole('tab', { name: 'Y' }));

      expect(view.asFragment()).toMatchSnapshot();
    });
  });

  it('lets user switch between X and Y settings tabs', async () => {
    createView();

    // should load X tab initially
    expect(screen.getByRole('tab', { selected: true })).toHaveTextContent('X');
    expect(screen.getByRole('tabpanel', { name: 'X' })).toBeVisible();
    expect(
      screen.queryByRole('tabpanel', { name: 'Y' })
    ).not.toBeInTheDocument();

    await user.click(screen.getByRole('tab', { name: 'Y' }));

    expect(screen.getByRole('tab', { selected: true })).toHaveTextContent('Y');
    expect(screen.getByRole('tabpanel', { name: 'Y' })).toBeVisible();
    expect(
      screen.queryByRole('tabpanel', { name: 'X' })
    ).not.toBeInTheDocument();
  });

  it('sets the correct values when plot variant changed from xy to timeseries', async () => {
    createView();

    await user.click(screen.getByRole('button', { name: 'Timeseries' }));

    expect(changeXAxis).toHaveBeenCalledWith('timestamp');
    expect(changeXAxisScale).toHaveBeenCalledWith('time');
    expect(changeXMinimum).toHaveBeenCalledWith(undefined);
    expect(changeXMaximum).toHaveBeenCalledWith(undefined);
  });

  it('sets the correct values when plot variant changed from timeseries to xy', async () => {
    props.XAxis = 'timestamp';
    props.XAxisScale = 'time';
    createView();

    await user.click(screen.getByRole('button', { name: 'XY' }));

    expect(changeXAxis).toHaveBeenCalledWith(undefined);
    expect(changeXAxisScale).toHaveBeenCalledWith('linear');
    expect(changeXMinimum).toHaveBeenCalledWith(undefined);
    expect(changeXMaximum).toHaveBeenCalledWith(undefined);
  });
});
