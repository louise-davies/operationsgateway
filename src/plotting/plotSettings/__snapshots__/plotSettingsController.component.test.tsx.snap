// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plot Settings component renders plot settings form correctly (x-axis tab selected) 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-column css-ffjoah-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <mock-plottitlefield
        data-testid="mock-plotTitleField"
      >
        plotTitle=undefined
changePlotTitle=undefined

      </mock-plottitlefield>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <mock-charttypebuttons
        data-testid="mock-chartTypeButtons"
      >
        plotType="scatter"
changePlotType=undefined

      </mock-charttypebuttons>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-1gsv261"
      >
        <div
          class="MuiTabs-root css-1yb2e4d-MuiTabs-root"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
            style="overflow: hidden; margin-bottom: 0px;"
          >
            <div
              aria-label="tabs"
              class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
              role="tablist"
            >
              <button
                aria-controls="X-tabpanel"
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-ejfd77-MuiButtonBase-root-MuiTab-root"
                id="X-tab"
                role="tab"
                tabindex="0"
                type="button"
              >
                X
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-controls="Y-tabpanel"
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-ejfd77-MuiButtonBase-root-MuiTab-root"
                id="Y-tab"
                role="tab"
                tabindex="-1"
                type="button"
              >
                Y
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
      </div>
      <div
        aria-labelledby="X-tab"
        id="X-tabpanel"
        role="tabpanel"
      >
        <div
          class="MuiBox-root css-0"
        >
          <mock-xaxistab
            data-testid="mock-xAxisTab"
          >
            allChannels=[
  {
    "systemName": "timestamp",
    "name": "Time",
    "type": "scalar",
    "path": "/system/timestamp"
  },
  {
    "systemName": "shotnum",
    "name": "Shot Number",
    "type": "scalar",
    "path": "/system/shotnum"
  },
  {
    "systemName": "activeArea",
    "name": "Active Area",
    "type": "scalar",
    "path": "/system/activeArea"
  },
  {
    "systemName": "activeExperiment",
    "name": "Active Experiment",
    "type": "scalar",
    "path": "/system/activeExperiment"
  },
  {
    "systemName": "test_1",
    "type": "scalar",
    "name": "Test 1",
    "precision": 4,
    "path": "/test_1"
  },
  {
    "systemName": "test_2",
    "type": "scalar",
    "precision": 2,
    "notation": "normal",
    "path": "/test_2"
  },
  {
    "systemName": "test_3",
    "type": "scalar",
    "precision": 2,
    "notation": "scientific",
    "path": "/test_3"
  }
]
XAxisScale="linear"
XAxis="test x-axis"
changeXAxis=undefined
changeXAxisScale=undefined
initialXMinimum=undefined
initialXMaximum=undefined
changeXMinimum=undefined
changeXMaximum=undefined

          </mock-xaxistab>
        </div>
      </div>
      <div
        aria-labelledby="Y-tab"
        hidden=""
        id="Y-tabpanel"
        role="tabpanel"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Plot Settings component renders plot settings form correctly (y-axis tab selected) 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-column css-ffjoah-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <mock-plottitlefield
        data-testid="mock-plotTitleField"
      >
        plotTitle=undefined
changePlotTitle=undefined

      </mock-plottitlefield>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <mock-charttypebuttons
        data-testid="mock-chartTypeButtons"
      >
        plotType="scatter"
changePlotType=undefined

      </mock-charttypebuttons>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-1gsv261"
      >
        <div
          class="MuiTabs-root css-1yb2e4d-MuiTabs-root"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
            style="overflow: hidden; margin-bottom: 0px;"
          >
            <div
              aria-label="tabs"
              class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
              role="tablist"
            >
              <button
                aria-controls="X-tabpanel"
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-ejfd77-MuiButtonBase-root-MuiTab-root"
                id="X-tab"
                role="tab"
                tabindex="-1"
                type="button"
              >
                X
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-controls="Y-tabpanel"
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-ejfd77-MuiButtonBase-root-MuiTab-root"
                id="Y-tab"
                role="tab"
                tabindex="0"
                type="button"
              >
                Y
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                >
                  <span
                    class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
                    style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
                  >
                    <span
                      class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
                    />
                  </span>
                </span>
              </button>
            </div>
            <span
              class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
      </div>
      <div
        aria-labelledby="X-tab"
        hidden=""
        id="X-tabpanel"
        role="tabpanel"
      />
      <div
        aria-labelledby="Y-tab"
        id="Y-tabpanel"
        role="tabpanel"
      >
        <div
          class="MuiBox-root css-0"
        >
          <mock-yaxistab
            data-testid="mock-yAxisTab"
          >
            selectedRecordTableChannels=[
  {
    "systemName": "timestamp",
    "channel_dtype": "scalar",
    "userFriendlyName": "Time"
  }
]
allChannels=[
  {
    "systemName": "timestamp",
    "name": "Time",
    "type": "scalar",
    "path": "/system/timestamp"
  },
  {
    "systemName": "shotnum",
    "name": "Shot Number",
    "type": "scalar",
    "path": "/system/shotnum"
  },
  {
    "systemName": "activeArea",
    "name": "Active Area",
    "type": "scalar",
    "path": "/system/activeArea"
  },
  {
    "systemName": "activeExperiment",
    "name": "Active Experiment",
    "type": "scalar",
    "path": "/system/activeExperiment"
  },
  {
    "systemName": "test_1",
    "type": "scalar",
    "name": "Test 1",
    "precision": 4,
    "path": "/test_1"
  },
  {
    "systemName": "test_2",
    "type": "scalar",
    "precision": 2,
    "notation": "normal",
    "path": "/test_2"
  },
  {
    "systemName": "test_3",
    "type": "scalar",
    "precision": 2,
    "notation": "scientific",
    "path": "/test_3"
  }
]
selectedPlotChannels=[]
changeSelectedPlotChannels=undefined
initialLeftYAxisMinimum=undefined
initialLeftYAxisMaximum=undefined
changeLeftYAxisMinimum=undefined
changeLeftYAxisMaximum=undefined
initialRightYAxisMinimum=undefined
initialRightYAxisMaximum=undefined
changeRightYAxisMinimum=undefined
changeRightYAxisMaximum=undefined
leftYAxisScale="logarithmic"
changeLeftYAxisScale=undefined
rightYAxisScale="linear"
changeRightYAxisScale=undefined
initialSelectedColours=undefined
initialRemainingColours=undefined
changeSelectedColours=undefined
changeRemainingColours=undefined

          </mock-yaxistab>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
